src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseSign1Message.cs

```diff
@@ -0,0 +1,283 @@
+            {
+                if (!unprotectedHeaders.TryGetEncodedValue(label, out encodedAlg))
+                {
+                    throw new CryptographicException(SR.Sign1SignAlgIsRequired);
+                }
+            }
+
+            return encodedAlg;
+        }
+
+        private static int? DecodeCoseAlgorithmHeader(ReadOnlyMemory<byte> encodedAlg)
+        {
+            var reader = new CborReader(encodedAlg);
+            CborReaderState state = reader.PeekState();
+
+            if (state == CborReaderState.NegativeInteger || state == CborReaderState.UnsignedInteger)
+            {
+                int alg = reader.ReadInt32();
+                KnownCoseAlgorithms.ThrowIfNotSupported(alg);
+                return reader.BytesRemaining == 0 ? alg : throw new CryptographicException(SR.Sign1VerifyAlgHeaderWasIncorrect);
+            }
+
+            if (state == CborReaderState.TextString)
+            {
+                int alg = KnownCoseAlgorithms.FromString(reader.ReadTextString());
+                return reader.BytesRemaining == 0 ? alg : throw new CryptographicException(SR.Sign1VerifyAlgHeaderWasIncorrect);
+            }
+
+            return null;
+        }
+
+        private static HashAlgorithmName GetHashAlgorithmFromCoseAlgorithmAndKeyType(int algorithm, KeyType keyType)
+        {
+            if (keyType == KeyType.ECDsa)
+            {
+                return algorithm switch
+                {
+                    KnownCoseAlgorithms.ES256 => HashAlgorithmName.SHA256,
+                    KnownCoseAlgorithms.ES384 => HashAlgorithmName.SHA384,
+                    KnownCoseAlgorithms.ES512 => HashAlgorithmName.SHA512,
+                    _ => throw new CryptographicException(SR.Format(SR.Sign1AlgDoesNotMatchWithTheOnesSupportedByTypeOfKey, algorithm, typeof(ECDsa)))
+                };
+            }
+            else
+            {
+                Debug.Assert(keyType == KeyType.RSA);
+                return algorithm switch
+                {
+                    KnownCoseAlgorithms.PS256 => HashAlgorithmName.SHA256,
+                    KnownCoseAlgorithms.PS384 => HashAlgorithmName.SHA384,
+                    KnownCoseAlgorithms.PS512 => HashAlgorithmName.SHA512,
+                    _ => throw new CryptographicException(SR.Format(SR.Sign1AlgDoesNotMatchWithTheOnesSupportedByTypeOfKey, algorithm, typeof(RSA)))
+                };
+            }
+        }
+
+        private static int GetCoseAlgorithmHeaderFromKeyTypeAndHashAlgorithm(KeyType keyType, HashAlgorithmName hashAlgorithm)
+            => keyType switch
+            {
+                KeyType.ECDsa => hashAlgorithm.Name switch
+                {
+                    KnownHashAlgorithms.SHA256 => KnownCoseAlgorithms.ES256,
+                    KnownHashAlgorithms.SHA384 => KnownCoseAlgorithms.ES384,
+                    KnownHashAlgorithms.SHA512 => KnownCoseAlgorithms.ES512,
```

Comment:
For a constant expression I'd just use `nameof(HashAlgorithmName.SHA512)` (and friends), and avoid the redeclared consts.